heron/packing/src/java/com/twitter/heron/packing/builder/PackingPlanBuilder.java

```diff
@@ -107,13 +112,80 @@ public PackingPlanBuilder addInstance(Integer containerId,
     return this;
   }
 
+  @SuppressWarnings("JavadocMethod")
+  /**
+   * Add an instance to the first container possible ranked by score.
+   * @return containerId of the container the instance was added to
+   * @throws com.twitter.heron.packing.ResourceExceededException if the instance could not be added
+   */
+  public int addInstance(Scorer<Container> scorer,
+                         InstanceId instanceId) throws ResourceExceededException {
+    List<Scorer<Container>> scorers = new LinkedList<>();
+    scorers.add(scorer);
+    return addInstance(scorers, instanceId);
+  }
+
+  @SuppressWarnings("JavadocMethod")
+  /**
+   * Add an instance to the first container possible ranked by score. If a scoring tie exists,
+   * uses the next scorer in the scorers list to break the tie.
+   * @return containerId of the container the instance was added to
+   * @throws com.twitter.heron.packing.ResourceExceededException if the instance could not be added
+   */
+  private int addInstance(List<Scorer<Container>> scorers, InstanceId instanceId)
+      throws ResourceExceededException {
+    initContainers();
+    for (Container container : sortContainers(scorers, this.containers.values())) {
+      try {
+        addInstance(container.getContainerId(), instanceId);
+        return container.getContainerId();
+      } catch (ResourceExceededException e) {
+        // ignore since we'll continue trying
+      }
+    }
+    //Not enough containers.
+    throw new ResourceExceededException(String.format(
+        "Insufficient resources to add instance %s to any of the %d containers.",
+        instanceId, this.containers.size()));
+  }
+
   public boolean removeInstance(Integer containerId, String componentName) {
     initContainers();
     Optional<PackingPlan.InstancePlan> instancePlan =
         containers.get(containerId).removeAnyInstanceOfComponent(componentName);
     return instancePlan.isPresent();
   }
 
+  @SuppressWarnings("JavadocMethod")
+  /**
+   * Remove an instance from the first container possible ranked by score.
+   * @return containerId of the container the instance was removed from
+   * @throws com.twitter.heron.spi.packing.PackingException if the instance could not be removed
+   */
+  public int removeInstance(Scorer<Container> scorer, String componentName) {
+    List<Scorer<Container>> scorers = new LinkedList<>();
+    scorers.add(scorer);
+    return removeInstance(scorers, componentName);
+  }
+
+  @SuppressWarnings("JavadocMethod")
+  /**
+   * Remove an instance from the first container possible ranked by score. If a scoring tie exists,
+   * uses the next scorer in the scorers list to break the tie.
+   * @return containerId of the container the instance was removed from
+   * @throws com.twitter.heron.spi.packing.PackingException if the instance could not be removed
+   */
+  public int removeInstance(List<Scorer<Container>> scorers, String componentName) {
```

Comment:
@billonahill I'm not sure if that is the same. In the previous example if we pass as input to the constructor container id 3 and containrId 5 then the search will stop at 5, right? Wehat we ideally want is to start from 3 and use this sequence: 4,5,1,2. Is that behavior captured?